src/time/Waveform.cpp

```diff
@@ -20,108 +21,84 @@ int Waveform::getInterpolationOrder() const
-  _storageIsInitialized  = true;
-  PRECICE_ASSERT(this->maxNumberOfStoredSamples() == storageSize);
-  PRECICE_ASSERT(this->valuesSize() == values.size());
-  for (int sampleIndex = 0; sampleIndex < maxNumberOfStoredSamples(); ++sampleIndex) {
-    this->storeAt(values, sampleIndex);
-  }
 }
 
-void Waveform::store(const Eigen::VectorXd &values)
+void Waveform::store(const Eigen::VectorXd &values, double normalizedDt)
 {
-  PRECICE_ASSERT(_storageIsInitialized);
-  int sampleIndex = 0;
-  this->storeAt(values, sampleIndex);
+  if (math::equals(_storage.maxStoredNormalizedDt(), 1.0)) { // reached end of window and trying to write new data from next window. Clearing window first.
+    bool keepZero = true;
+    _storage.clear(keepZero);
+  }
+  PRECICE_ASSERT(values.size() == _storage.nDofs());
+  _storage.setValueAtTime(normalizedDt, values);
 }
 
-void Waveform::storeAt(const Eigen::VectorXd values, int sampleIndex)
+// helper function to compute x(t) from given data (x0,t0), (x1,t1), ..., (xn,tn) via B-spline interpolation (implemented using Eigen).
+Eigen::VectorXd bSplineInterpolationAt(double t, Eigen::VectorXd ts, Eigen::MatrixXd xs, int splineDegree)
 {
-  PRECICE_ASSERT(_timeWindowsStorage.cols() > sampleIndex, maxNumberOfStoredSamples(), sampleIndex);
-  PRECICE_ASSERT(values.size() == this->valuesSize(), values.size(), this->valuesSize());
-  this->_timeWindowsStorage.col(sampleIndex) = values;
+  // organize data in columns. Each column represents one sample in time.
+  PRECICE_ASSERT(xs.cols() == ts.size());
+  const int ndofs = xs.rows(); // number of dofs. Each dof needs it's own interpolant.
+
+  Eigen::VectorXd interpolated(ndofs);
+
+  const int splineDimension = 1;
+
+  for (int i = 0; i < ndofs; i++) {
+    auto spline     = Eigen::SplineFitting<Eigen::Spline<double, splineDimension>>::Interpolate(xs.row(i), splineDegree, ts);
+    interpolated[i] = spline(t)[0]; // get component of spline associated with xs.row(i)
+  }
+
+  return interpolated;
 }
 
 Eigen::VectorXd Waveform::sample(double normalizedDt)
 {
-  PRECICE_ASSERT(_storageIsInitialized);
-  PRECICE_ASSERT(normalizedDt >= 0, "Sampling outside of valid range!");
-  PRECICE_ASSERT(normalizedDt <= 1, "Sampling outside of valid range!");
+  const int usedOrder = computeUsedOrder(_interpolationOrder, _storage.nTimes());
 
-  const int usedOrder = computeUsedOrder(_interpolationOrder, _numberOfStoredSamples);
+  PRECICE_ASSERT(math::equals(this->_storage.maxStoredNormalizedDt(), 1.0), this->_storage.maxStoredNormalizedDt()); // sampling is only allowed, if a window is complete.
 
-  if (usedOrder == 0) {
+  // @TODO: Improve efficiency: Check whether key = normalizedDt is in _storage. If yes, just get value and return. No need for interpolation.
+
+  if (_interpolationOrder == 0) {
+    // @TODO: Remove constant interpolation in preCICE v3.0? Usecase is unclear and does not generalize well with BSpline interpolation. It's also not 100% clear what to do at the jump.
```

Comment:
@uekerman what do you think about this point? Should we just remove constant interpolation in 3.0.0 and make linear interpolation the new default? It should still behave the same way as before and I don't see a real benefit providing constant interpolation (maybe for academic reasons but that's it).

Reply:
It's also a bit undocumented right now which "constant interpolation" we mean, constant over the complete time window or constant per timestep. I guess the later would be more consistent with what we do for order 1 and higher.
I think that the academic use case could be interesting in the long run. To quickly test things (whether the overall order drops or not).
So if it is not overly complicated to keep that case, I would vote for the "constant per timestep".

Reply:
It's no big problem to keep the constant interpolation. So let's keep it. I think the academic argument makes sense. We can still remove it at some point in time when the academic case is not relevant anymore - currently it definitely is for my thesis ;)

On the question interpolation on time step vs. interpolation on window: I would not solve this problem here. The waveform class currently just uses all the data it has. It will use the "per-window-interpolation", if there is only data at the beginning and at the end of the window. If there is more data: It will use all the data that is available and do piecewise constant, linear ... interpolation ("per-time-step-interpolation"). I will add a few more lines of documentation here.

Note: In the end we can still use the "per-window-interpolation", if we just provide data at the beginning and at the end of the window and ignore all the data from subcycling (basically what's happening in preCICE v2). But this is mainly a matter of configuration. Question here: Do we even want to keep this option? If I want to ignore data, why do I even write it in the first place? Again: The only reason I see for still offering the v2 subcycling is academic. But let's discuss this in the "big" PR, where we will also deal with configuration.

Reply:
I added some more documentation in [53fda27](https://github.com/precice/precice/pull/1422/commits/53fda271c4f6a78808dad6874098a446bba28038). I hope it is clear now.